File: app/src/main/java/pers/victor/androiddemos/util/CustomBehavior.java
package pers.victor.androiddemos.util;

import android.animation.Animator;
import android.content.Context;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.View;

import pers.victor.androiddemos.AndroidDemos;

/**
 * Created by Victor on 2017/4/1. (ง •̀_•́)ง
 */

public class CustomBehavior extends FloatingActionButton.Behavior {
    private boolean isAnim = false;
    private boolean isHide = false;
    private float y = 0f;

    public CustomBehavior(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public boolean onStartNestedScroll(CoordinatorLayout coordinatorLayout, FloatingActionButton child, View directTargetChild, View target, int nestedScrollAxes) {
        if (y == 0f) {
            y = child.getY();
        }
        return nestedScrollAxes == ViewCompat.SCROLL_AXIS_VERTICAL;
    }


    @Override
    public void onNestedScroll(CoordinatorLayout coordinatorLayout, FloatingActionButton child, View target, int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed) {
        if (isAnim) {
            return;
        }
        if (dyConsumed > 0 && !isHide) {
            hide(child);
        }
        if (dyConsumed < 0 && isHide) {
            show(child);
        }
    }

    private void hide(FloatingActionButton child) {
        isAnim = true;
        isHide = true;
        SimpleAnimator.create(child)
                .values(y, (float) AndroidDemos.screenHeight)
                .property("y")
                .duration(500)
                .listener(new SimpleAnimator.AnimatorListener() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        isAnim = false;
                    }
                })
                .go();
    }

    private void show(FloatingActionButton child) {
        isAnim = true;
        isHide = false;
        SimpleAnimator.create(child)
                .values((float) AndroidDemos.screenHeight, y)
                .property("y")
                .duration(500)
                .listener(new SimpleAnimator.AnimatorListener() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        isAnim = false;
                    }
                })
                .go();
    }
}

File: app/src/main/kotlin/demos/util/behavior/CustomBehavior.kt
package demos.util.behavior

import android.animation.Animator
import android.content.Context
import android.support.design.widget.CoordinatorLayout
import android.support.design.widget.FloatingActionButton
import android.support.v4.view.ViewCompat
import android.util.AttributeSet
import android.view.View
import demos.AndroidDemos
import demos.util.SimpleAnimator

/**
 * Author : victor
 * Time : 16-9-20 10:10
 */
class CustomBehavior(context: Context, attributeSet: AttributeSet) : FloatingActionButton.Behavior(context, attributeSet) {
    private var isAnim = false
    private var isHide = false
    private var y = 0f

    override fun onStartNestedScroll(coordinatorLayout: CoordinatorLayout?, child: FloatingActionButton, directTargetChild: View?, target: View?, nestedScrollAxes: Int): Boolean {
        if (y == 0f) {
            y = child.y
        }
        return nestedScrollAxes == ViewCompat.SCROLL_AXIS_VERTICAL
    }

    override fun onNestedScroll(coordinatorLayout: CoordinatorLayout, child: FloatingActionButton, target: View, dxConsumed: Int, dyConsumed: Int, dxUnconsumed: Int, dyUnconsumed: Int) {
        if (isAnim) {
            return
        }
        if (dyConsumed > 0 && !isHide) {
            hide(child)
        }
        if (dyConsumed < 0 && isHide) {
            show(child)
        }
    }

    private fun hide(child: FloatingActionButton) {
        isAnim = true
        isHide = true
        SimpleAnimator(child).values(y, AndroidDemos.screenHeight).property("y").duration(500).listener(object : SimpleAnimator.AnimatorListener() {
            override fun onAnimationEnd(animation: Animator?) {
                isAnim = false
            }
        }).go()
    }

    private fun show(child: FloatingActionButton) {
        isAnim = true
        isHide = false
        SimpleAnimator(child).values(AndroidDemos.screenHeight, y).property("y").duration(500).listener(object : SimpleAnimator.AnimatorListener() {
            override fun onAnimationEnd(animation: Animator?) {
                isAnim = false
            }
        }).go()
    }

}